raft/go.mod

```diff
@@ -12,17 +12,23 @@ require (
 	go.etcd.io/etcd/client/pkg/v3 v3.6.0-alpha.0
 )
 
+require (
```

Comment:
I think we should not put raft fuzz tests in the same package. There have been already requests to stop polluting package with test dependencies https://github.com/etcd-io/etcd/issues/14513. This will do it even more. 

I would propose to create rafttest package in this PR to avoid adding more dependencies to raft.
Would be good to get opinion from @tbg and other people working on raft @nvanbenschoten @pavelkalinnikov

Reply:
Maybe there are some guidelines I'm not aware of, but the "go way" is to write unit tests in the same package. Unit tests use helpers such as datadriven, require, fuzzing, etc and so these would be in the same package's tests. I would be very hesitant to bend over backwards and test in unusual ways just because we're worried about `go.mod` hygiene.

Testing dependencies will (should) be pruned by anyone importing the package[^1], i.e. won't actually be pulled.
For example, we use `datadriven` in tests (and it's even used in the main `rafttest` package), but because `raft/v3` doesn't reference `rafttest`, it's not a dependency:

```
go list -deps | grep datadriven
```

The same will be the case here, so I don't see an issue.

[^1]: https://go.dev/ref/mod#graph-pruning

Reply:
You probably mean creating a separate module? I've seen a few projects using this technique: https://dev.to/wallyqs/introducing-gotest-mod-18a1. Although I think it might be unnecessary with Go [module graph pruning](https://go.dev/ref/mod#graph-pruning): https://stackoverflow.com/questions/64071364/best-way-to-use-test-dependencies-in-go-but-prevent-export-them.

Reply:
Sorry, didn't see Tobi's comment. But we essentially agree.

Reply:
+1 on not adding a lot of transitive deps for fuzzing. with fuzz support being added directly to go, I would expect to see fuzzing-related dependencies decreasing, not increasing

Reply:
Indeed the `datadriven` will not be included in the the module graph of a module which depends on `raft` because it's pruned as long as the modules specifies go 1.17 or higher. The reason of https://github.com/etcd-io/etcd/issues/14513 is the etcd 3.5.x specifies go 1.16.  So adding the dependency "github.com/AdaLogics/go-fuzz-header" will not pollute the module graph of Kubernetes any more, because both etcd `main` (3.6) and Kubernetes `main` already specify go 1.19.

So the focus for now is should etcd accept adding the new dependency into `raft`? Personally I think it might be OK, as long as the dependency is well maintained.

Reply:
I still think the project should avoid taking on additional dependencies like this, but if they don't show up in `go mod graph`, I object less strenuously.